padl/transforms.py

```diff
@@ -754,84 +793,60 @@ def pd_device(self) -> str:
         """Return the device ("cpu" / "cuda") the transform is on."""
         return self._pd_device
 
+    @pd_device.setter
+    def pd_device(self, device: str):
+        self._pd_device = device
+
     @property
     def pd_preprocess(self) -> "Transform":
         """The preprocessing part of the transform. The device must be propagated from self."""
-        pre = self._pd_get_stages()[0]
-        pre.pd_to(self.pd_device)
+        pre = self._pd_get_stages[0]
+        pre.pd_device = self.pd_device
         return pre
 
     @property
     def pd_forward(self) -> "Transform":
         """The forward part of the transform (that what's typically done on the GPU).
         The device must be propagated from self."""
-        forward = self._pd_get_stages()[1]
-        forward.pd_to(self.pd_device)
+        forward = self._pd_get_stages[1]
+        forward.pd_device = self.pd_device
         return forward
 
     @property
     def pd_postprocess(self) -> "Transform":
         """The postprocessing part of the transform. The device must be propagated from self."""
-        post = self._pd_get_stages()[2]
-        post.pd_to(self.pd_device)
+        post = self._pd_get_stages[2]
+        post.pd_device = self.pd_device
         return post
 
     def pd_to(self, device: str) -> "Transform":
         """Set the transform's device to *device*.
 
         :param device: Device to set the transform to {'cpu', 'cuda', 'cuda:N'}.
         """
-        self._pd_device = device
+        self.pd_device = device
         for layer in self.pd_layers:
             layer.to(device)
         return self
 
     @property
+    @lru_cache(maxsize=128)
     def pd_layers(self) -> List[torch.nn.Module]:
         """Get a list with all pytorch layers in the transform (including layers in sub-transforms).
         """
-        layers = []
-        for subtrans in self._pd_all_transforms():
-            if isinstance(subtrans, torch.nn.Module):
-                layers.append(subtrans)
-        return layers
+        if self._pd_layers is None:
```

Comment:
as this is being cached now, do we need to do this?

Reply:
No we do not, I'll change this as well.